src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, can_reshape, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (is_optimized_multiplication()) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+  int in1_op = in1->Opcode();
+
+  bool matched = false;
+  jlong multiplier;
+
+  // Convert (a + a) + a to 3 * a
+  // Look for LHS pattern: AddNode(a, a)
+  if (in1_op == Op_Add(bt) && in1->in(1) == in1->in(2)) {
+    if (in1->in(1) != in2) {
+      return nullptr;
+    }
+
+    multiplier = 2;
+    matched = true;
+  }
+
+  // Convert (a << CON) + a to a * ((1 << CON) + 1))
+  // Look for LHS pattern: LShiftNode(a, CON)
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (in1_op == Op_LShift(bt) && in1->in(2)->is_Con()) {
+    Node* base = in1->in(1);
+    if (base != in2) {
+      return nullptr;
+    }
+
+    Node* con = in1->in(2);
+    BasicType con_bt = phase->type(con)->basic_type();
+    if (con_bt == T_VOID) { // const could potentially be void type
```

Comment:
Does that happen when con is top? In that case I think it's better to use `con->is_top()`. Or maybe `find_integer_as_long` if there's a value we know for sure that con can't take (it can't be negative, right?).

Reply:
Yes it is better. Will use`is_top()` instead.

> if there's a value we know for sure that con can't take (it can't be negative, right?)

I thought about `find_integer_as_long(-1)`, but it could be anything, even negative. This transformation accounts for example `a*-43 + a` => `a*-42`.